### PR TITLE
Only write options after onboarding

### PR DESCRIPTION
**Pull request by johnny9**
[Windows](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/284)
[Intel macOS](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/284)
[Apple Silicon macOS](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/284)
[ARM64 Android](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/unsecure_android_apk.zip?branch=pull/284)

### CODE REVIEW

**Comment by hebasto** [comment]
Needs rebase.

**Comment by johnny9** [comment]
update from[ 45e3262 ](https://github.com/bitcoin-core/gui-qml/commit/45e326291780c5832950518fbdf305defb53458c)to 08be85b:
 - rebased over main

**Comment by thebagmaster** [comment]
ACK [08be85b](https://github.com/bitcoin-core/gui-qml/commit/08be85b55feb5fcb6b2a04c669fd91c74fdabbf5)
I tested these on Ubuntu 22.04.2 and it does what it says on the tin. 
Understood that we may need to buffer the settings writing to file in on-boarding; especially if the storage location is not set and may change.
Maybe we should do our best to save what we can of what the user configured?
Are there situations where we buffer and different ones that we save?

**Comment by hebasto** [comment]
Please rebase.

https://discord.com/channels/903125802726596648/903126039084015638/1146922061449666602

**Comment by johnny9** [comment]
Update from[ 08be85b ](https://github.com/bitcoin-core/gui-qml/commit/08be85b55feb5fcb6b2a04c669fd91c74fdabbf5)to 51df97d
 - rebased with main

**Review by jarolrod — COMMENTED** [review]
Concept ACK

**Comment by D33r-Gee** [comment]
Built the apk however it is not behaving as expected, it is writing the `settings.json` file before finishing the onboarding process.

Also the apk launches on `mainnet` although the “-signet” is set in the AndroidManifest.xml file.

**Comment by D33r-Gee** [comment]
tACK on Android armv7 everything worked as expected

The `settings.json` was created after the onboarding with the following values:

`{
    "prune": 2861
}`

**Comment by D33r-Gee** [comment]
ACK 51df97d142558919e9181696359debfb98e94b0c

**Review by D33r-Gee — COMMENTED** [review]
ACK 51df97d142558919e9181696359debfb98e94b0c

**Review by pablomartin4btc — COMMENTED** [review]
Concept ACK, I'll test soon. It looks that there are some unaddressed feedback.

**Review by pablomartin4btc — COMMENTED** [review]
tACK 51df97d142558919e9181696359debfb98e94b0c

Tested in both `master`(/ `main`) and this PRs branch: changing all settings touched by this branch's code change and verify the changes were saved accordingly.

Things I noticed during testing:

In `main`:
- If the user doesn't change anything `setting.json` file will look like:

```
{
    "prune": 1907
}
```

- If the user disables any settings `setting.json` file will look like:

```
{
    "prune": 0
    "listen": false,
    "natpmp": false,
    "server": false,
    "upnp": false

}
```

In branch `johnny9:write_after_onboard`:
- If the user doesn't change anything `setting.json` file will look like:

```
{
    "dbcache": 16000,
    "listen": true,
    "natpmp": true,
    "prune": 1907,
    "server": true,
    "upnp": true
}
```

- If the user disables any settings `setting.json` file will look like:

```
{
    "dbcache": 16000,
}
```


<details>

<summary>After the user makes any changes on the settings mentioned above next time the user starts `bitcoin-qt` up again it looks like the UI doesn't match the configuration file (`settings.json`), is this expected?</summary>


- Using `main`:

Screenshot from 2024-03-12 15-55-36

- Using this branch:

Screenshot from 2024-03-12 16-03-51


</details>

**Comment by johnny9** [comment]
Update  51df97d to 61f8e6c 
 - Rebased with master
 - Initialized all uninitialized OptionsQmlModel properties

**Comment by D33r-Gee** [comment]
tACK https://github.com/bitcoin-core/gui-qml/commit/61f8e6c5fcb0c6c31a6283c3b7fdf6b9f82eba5e

On Ubuntu 22.04 and Android (armv7)

Works as expected, settings.json shows defaults:

```
{
    "listen": true,
    "prune": 1907
}
```

**Review by pablomartin4btc — APPROVED** [review]
tACK 61f8e6c5fcb0c6c31a6283c3b7fdf6b9f82eba5e

It behaves as expected.

`dbcache` setting is not persisted anymore (`"dbcache": 16000,`), is this correct?

**Comment by johnny9** [comment]
> `dbcache` setting is not persisted anymore (`"dbcache": 16000,`), is this correct?

It should only persist if you set it to a value other than nDefaultDbCache (450)

**Comment by pablomartin4btc** [comment]
> > `dbcache` setting is not persisted anymore (`"dbcache": 16000,`), is this correct?
> 
> It should only persist if you set it to a value other than nDefaultDbCache (450)

Perfect, it works.

**Comment by johnny9** [comment]
Update  from 61f8e6c to 4f557b0 
 - rebased with main

**Review by D33r-Gee — COMMENTED** [review]
tACK 9e569074a1735976ffb324dfb5531c24591f3252

lgtm, works as expected on both Ubuntu 22.04 and Android (armv7)

played around with different settings and they registered properly both in the settings.json and in the debug.log

**Review by pablomartin4btc — APPROVED** [review]
re-ACK 9e569074a1735976ffb324dfb5531c24591f3252

**Review by MarnixCroes — COMMENTED** [review]
9e569074a1735976ffb324dfb5531c24591f3252
writing of the settings works, but it always additionally creates a `settings.json.bak` file?

File: src/qml/models/options_model.cpp [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include <common/settings.h>
 #include <common/system.h>
 #include <interfaces/node.h>
+#include <mapport.h>
 #include <qt/guiconstants.h>
 #include <qt/guiutil.h>
 #include <qt/optionsmodel.h>
```
Comment (pablomartin4btc) [inline src/qml/models/options_model.cpp:14]: Do we need this?
```suggestion
```

Comment (johnny9) [reply src/qml/models/options_model.cpp:14]: It is needed for PruneMiBtoGB i believe

File: src/qml/models/options_model.cpp [hunk 2]
```diff
@@ -136,4 +162,31 @@ void OptionsQmlModel::setCustomDataDirArgs(QString path)
...
+        m_node.updateRwSetting("par", m_script_threads);
+    }
+    if (m_server) {
+        m_node.updateRwSetting("server", m_server);
+    }
+    if (m_upnp) {
+        m_node.updateRwSetting("upnp", m_upnp);
+    }
+    m_onboarded = true;
+}
```
[the remark this answers is not in the document]
Comment (johnny9) [reply src/qml/models/options_model.cpp:192]: already has a newline character at the end of the file

**Review by D33r-Gee — COMMENTED** [review]
re ACK 9e569074a1735976ffb324dfb5531c24591f3252 on Ubuntu 22.04 works as expected

**Pull request merged**